### DISCUSSION TITLE
Usage in Windows / MSVC / Unreal Engine Contexts

### DISCUSSION THREAD

**Discussion by MrMoose** [googleapis/google-cloud-cpp#6202] [General]
Hello everyone,

I hope this is a suitable place to spark a discussion about ease of integration of this library in non-linux environments, specifically when it gets a bit "weirder", such as within the Unreal engine. Please take this as constructive criticism if possible. It is meant as such.

Let me illustrate my journey here a little. I come from a 20 year background in C++ and have quite a bit of experience building various Google libs such as protobuf, gRPC and others on Linux and Windows. Over the last years I have seen and tackled many of the challenges that come with it. This time around, I'm working on a new (Open Source) plugin for the Unreal Engine which is supposed to make use of google-cloud-cpp. Which basically requires me to produce a MSVC build of the library plus all dependencies for linkage within a DLL. It took a long time but finally I have a build and can integrate into my Unreal plugin.

Which brings me to my first topic. Those dependencies and building a Windows version of this library really are a lot. It took many days and without my prior experience with gRPC I doubt I could have done it. If you excuse me comparing this to the AWS SDK for C++, which I have done the same thing with, it was a matter of an hour or so. Clone the repo, run CMake, build, go. Easy. I argue that the build dependencies involved here are either too great or should be hidden from a user's build system enough for people to create a Windows build without having to resort to building all the dependencies as well. If only to be competitive with the competition's SDK here.

So, moving on. To be able to compile I needed a workaround for two compile errors that appeared as soon as I include anything from this library. I needed to disable compile [errors C4582 C4583 as seen here]( https://social.msdn.microsoft.com/Forums/Lync/en-US/c33e7d32-6e90-42ce-8ceb-2664f2cfa036/compiler-errors-c4582-and-c4583-constructordestructor-is-not-implicitly-called?forum=vcgeneral).
A similar use case to your usage in `gc::StatusOr<>`. Apparently MSVC won't have it and Unreal declared this an error rather than a warning, which is why I can disable them using:

```C++
#pragma warning(push)
#pragma warning(disable:4583)
#pragma warning(disable:4582)
#include "google/cloud/storage/client.h"
#pragma warning(pop)
```

With things finally building, I have started to implement a simple write operation to Storage. Very much like in your examples. It compiles and seems to be able to create the client object but unfortunately the actual write() operation fails. No matter what I try, I'm always getting an exception of type `absl::lts_2020_09_23::bad_optional_access` when I dereference the StatusOr. I am suspecting this is directly related with above disabled compiler warnings.

Which brings me to my next topic: Let's call it "integration friendlyness". C++ is an incredibly diverse cosmos with a huge number of flavors, language versions and so on. Unreal being by far not the freakiest I've come across.

Starting with the question why the exception is raised in the first place. In the CMake settings this lib exposes a setting to enable or disable exceptions. I have them enabled as you state you would std::abort() otherwise anyway and catch them in Unreal code. But is the choice between using exceptions and std::abort() really a choice? It feels more like "use exceptions or crash" to me. Specifically within an environment where I can' influence std::abort. I would consider the choice between exceptions and some kind of response code or outcome a better choice.

Another question is the usage of types like StatusOr. I know, I'm an old school dude but in my view, it is good for a client library to be easy to integrate. Which means, it's better to do things in a 'usual' or common way, even though they might not be as fancy as they could be. In order for as many potential users as possible to look at an example and say, OK, I see.
Can't a factory create that client as a shared ptr which might be null? Otherwise, you have exceptions on anyway, you could just throw when the client can't be created. 
Don't get me wrong, I appreciate new ways of doing things but they ought to be hidden in the impl of a client library, not forcibly exposed to the end users. They might be in an environment where they can easily deal with that, or not. It may even turn out to be a showstopper, as I am starting to fear it is for me.
In my view, the simpler and more common language features an API interface uses, the better. The higher the possibility that integration into any environment is possible.

So, I hope you didn't take this as a rant. I really just try to contribute some feedback about this library. Perhaps someone may find this useful.

Cheers,

Moose

**Comment by coryan** [comment]
> Hello everyone,

Hello!  Thanks for joining the discussion.

> Let me illustrate my journey here a little. I come from a 20 year background in C++ and have quite a bit of experience building various Google libs such as protobuf, gRPC and others on Linux and Windows. 

Nice.

> Over the last years I have seen and tackled many of the challenges that come with it. This time around, I'm working on a new (Open Source) plugin for the Unreal Engine which is supposed to make use of google-cloud-cpp. Which basically requires me to produce a MSVC build of the library plus all dependencies for linkage within a DLL. It took a long time but finally I have a build and can integrate into my Unreal plugin.

Which portions of `google-cloud-cpp` are you interested in using?  If (and based on the discussion below this sounds plausible) you only want GCS, then you could drop the dependencies on gRPC and protobuf.  The magic incantation is:

```
cmake -DGOOGLE_CLOUD_CPP_ENABLE=storage -S . -B ...
```

> 
> Which brings me to my first topic. Those dependencies and building a Windows version of this library really are a lot. It took many days and without my prior experience with gRPC I doubt I could have done it. If you excuse me comparing this to the AWS SDK for C++, which I have done the same thing with, it was a matter of an hour or so. Clone the repo, run CMake, build, go. Easy. I argue that the build dependencies involved here are either too great or should be hidden from a user's build system enough for people to create a Windows build without having to resort to building all the dependencies as well. If only to be competitive with the competition's SDK here.

We know this is a problem, and it is something we are constantly working on.  The next release might help, building all the dependencies becomes:

```
cd $HOME
git clone https://github.com/Microsoft/vcpkg
git clone https://github.com/googleapis/google-cloud-cpp
cd google-cloud-cpp
cmake -DCMAKE_TOOLCHAIN_FILE=$HOME/vcpkg/scripts/buildsystems/vcpkg.cmake -S . -B ...
```

That should download and compile all the dependencies for you.  It will not solve the DLL problems, for the reasons in #5849.

> 
> So, moving on. To be able to compile I needed a workaround for two compile errors that appeared as soon as I include anything from this library. I needed to disable compile [errors C4582 C4583 as seen here](https://social.msdn.microsoft.com/Forums/Lync/en-US/c33e7d32-6e90-42ce-8ceb-2664f2cfa036/compiler-errors-c4582-and-c4583-constructordestructor-is-not-implicitly-called?forum=vcgeneral).
> A similar use case to your usage in `gc::StatusOr<>`. Apparently MSVC won't have it and Unreal declared this an error rather than a warning, which is why I can disable them using:

I am slightly confused, we compile with MSVC (2017 and 2019) after each PR.  We also compile with `/WX` `/W4`... Can you let us know what version of MSVC are you compiling with?  Maybe we need to add that version to our build matrix, or maybe those warnings are disabled for some reason.

> 
> ```c++
> #pragma warning(push)
> #pragma warning(disable:4583)
> #pragma warning(disable:4582)
> #include "google/cloud/storage/client.h"
> #pragma warning(pop)
> ```

Ack.  I cannot find much documentation about those warnings beyond "introduced in MSVC 2015 RTM":

https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warnings-by-compiler-version?view=msvc-160#warnings-introduced-in-visual-studio-2015-rtm-compiler-version-1900230260

> With things finally building, I have started to implement a simple write operation to Storage. Very much like in your examples. It compiles and seems to be able to create the client object but unfortunately the actual write() operation fails.

Okay... Can you try a simpler operation first, like `GetObjectMetadata` or `GetBucketMetadata`?

> No matter what I try, I'm always getting an exception of type `absl::lts_2020_09_23::bad_optional_access` when I dereference the StatusOr. I am suspecting this is directly related with above disabled compiler warnings.

That is possible.  It could also be a runtime problem.  How are you dereferencing the `StatusOr<>`?  Just `operator*`, or `.value()`? Do you check the status in the `StatusOr<>`?  Can you print it?

Were you able to compile and test any of the examples with your setup?

> Which brings me to my next topic: Let's call it "integration friendlyness". C++ is an incredibly diverse cosmos with a huge number of flavors, language versions and so on. Unreal being by far not the freakiest I've come across.

We should share a beverage some day and complain about C++ freaky environments.

> 
> Starting with the question why the exception is raised in the first place. In the CMake settings this lib exposes a setting to enable or disable exceptions. I have them enabled as you state you would std::abort() otherwise anyway and catch them in Unreal code. But is the choice between using exceptions and std::abort() really a choice? It feels more like "use exceptions or crash" to me. Specifically within an environment where I can' influence std::abort. I would consider the choice between exceptions and some kind of response code or outcome a better choice.

Well, that is what we offer, a `StatusOr<>` *is* an outcome-type class.  There are a few places where you may ask us to raise an exception or error (e.g. `StatusOr<>::value()`, the iostreams when setting some exception flags) and **if** you do that **and** you disabled exceptions at compile time, **then** we call `std::abort()`.  Not like we have much choice.

> Can't a factory create that client as a shared ptr which might be null? Otherwise, you have exceptions on anyway, you could just throw when the client can't be created.

Frankly, I would prefer to just throw, but as you noted earlier some folks disable exceptions. Returning `nullptr` would indicate that *something* failed, but not *what*, returning `StatusOr<std::shared_ptr<T>>` is err .. not the most elegant thing ... but at least you get to know what failed.

I must confess that I do not know at this point if using exceptions is a problem for you or not.  It seems that you can use them.  If I am correct please consider using `StatusOr<T>::value()`, as in:

```cc
StatusOr<Foo> G();

void F() {
  Foo v = G().value();
}
```

> So, I hope you didn't take this as a rant. I really just try to contribute some feedback about this library. Perhaps someone may find this useful.

I took it as "Okay, we still have work to do in (a) making this easier to compile / integrate, (b) easier to use in larger frameworks like Unreal, and (c) document how `StatusOr<>` is supposed to be used".

**Reply by MrMoose** [reply]
> Hello! Thanks for joining the discussion.

Whoa, thanks for your unexpectedly prompt response. Let me go through some of the point you make.

> Which portions of `google-cloud-cpp` are you interested in using? If (and based on the discussion below this sounds plausible) you only want GCS, then you could drop the dependencies on gRPC and protobuf. The magic incantation is:
> 
> ```
> cmake -DGOOGLE_CLOUD_CPP_ENABLE=storage -S . -B ...
> ```

Sadly, this may not be an option. What I require is (at least now) storage, pubsub and ideally logging even though I came to realize that logging is not really available right now. Looks like you have it planned though. Once I understood this I changed my plan to integrate storage first, then pubsub and then wait for logging to be integrated.

> > [Build dependencies]
> 
> We know this is a problem, and it is something we are constantly working on. The next release might help, building all the dependencies becomes:
> 
> ```
> cd $HOME
> git clone https://github.com/Microsoft/vcpkg
> git clone https://github.com/googleapis/google-cloud-cpp
> cd google-cloud-cpp
> cmake -DCMAKE_TOOLCHAIN_FILE=$HOME/vcpkg/scripts/buildsystems/vcpkg.cmake -S . -B ...
> ```

Well that looks interesting. Just bear in mind that package managers may not be the silver bullet. I shy away from them when I can because they are forced to make a lot of assumptions about your environment.
Unreal is a good example: They build everything with /MD. Which is great, but unexpectedly, this is also the case for debug builds (as apposed to /MDd). I call that "fake debug". No matter what and how you build, you're always using release runtime. Their debug builds are actually release builds with compiler optimization turned off and debug info turned on. Works just as well as normal debug builds except you don't need debug library dependencies and you can freely mix your builds. I found that to be so great that I use this everywhere instead of bothering with "real" debug.
Package managers though have no idea about this and often will require twin builds of release and debug because it is written. That's just one example. If your solution you describe will produce a build with vcpkg I will certainly give it a try.

> That should download and compile all the dependencies for you. It will not solve the DLL problems, for the reasons in #5849.

Ahh, that looks familiar. I used to have that problem with protobuf on Windows as well. As far as I knw there was a protoc option that allowed you to inject declspec decorators of your choice into the generated headers. Isn't that what you need here?

> I am slightly confused, we compile with MSVC (2017 and 2019) after each PR. We also compile with `/WX` `/W4`... Can you let us know what version of MSVC are you compiling with? Maybe we need to add that version to our build matrix, or maybe those warnings are disabled for some reason.

This is Visual Studio 2019, latest patch. Though I don't believe it would occur in a regular VS solution. Unreal only uses VS as some sort of an editor but the actual build system used underneath is UBT (Unreal Build Tool) which does a lot of magick on top and then calls the MSVC compiler. Somewhere in there is probably the reason for this.

> > ```c++
> > #pragma warning(push)
> > #pragma warning(disable:4583)
> > #pragma warning(disable:4582)
> > #include "google/cloud/storage/client.h"
> > #pragma warning(pop)
> > ```
> 
> Ack. I cannot find much documentation about those warnings beyond "introduced in MSVC 2015 RTM":
> 
> https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warnings-by-compiler-version?view=msvc-160#warnings-introduced-in-visual-studio-2015-rtm-compiler-version-1900230260

There's an interesting resource here: https://github.com/akrzemi1/Optional/issues/57

It looks to me as if this is not unheard of when using optional types in this context. How I could fix this is over my head though. I am currently toying around with setting various language options such as C++14. The fix above only worked when using 17. Using 14 so far I only got a compile error "\GoogleCloud_SDK\include\absl/meta/type_traits.h(298): error C4596: 'value': illegal qualified name in member declaration" which I couldn't fix yet.

> Okay... Can you try a simpler operation first, like `GetObjectMetadata` or `GetBucketMetadata`?

I will try that and update here. Alas, I doubt it as the problem seems to occur when I dereference the StatusOr, before the actual call happens. At least this is what the debugger says.

> > No matter what I try, I'm always getting an exception of type `absl::lts_2020_09_23::bad_optional_access` when I dereference the StatusOr. I am suspecting this is directly related with above disabled compiler warnings.
> 
> That is possible. It could also be a runtime problem. How are you dereferencing the `StatusOr<>`? Just `operator*`, or `.value()`? Do you check the status in the `StatusOr<>`? Can you print it?

I do check it. Here's a snippet of my code:

```C++
gc::StatusOr<gcs::Client> client = gcs::Client::CreateDefaultClient();
if (!client) {
   // handle error using client.status().message()
   return;
}

gc::StatusOr<gcs::ObjectMetadata> object_metadata = client->InsertObject(bucket_name, object_key, mydata, gcs::ContentType(content_type));
			
// never get here. absl Exception is thrown
```

My original code used WriteObject() and then write a buffer of about 16 bytes into it but there it's even weirder and I get a bad_alloc at runtime.

> Were you able to compile and test any of the examples with your setup?

No, but I haven't tried. I fear this will work outside Unreal and send me on a wild goose chase nowhere but you're right. I should probably give it a go to verify that the build itself is OK.

> We should share a beverage some day and complain about C++ freaky environments.

... and invite cross compile users ;-)

> Frankly, I would prefer to just throw, but as you noted earlier some folks disable exceptions. Returning `nullptr` would indicate that _something_ failed, but not _what_, returning `StatusOr<std::shared_ptr<T>>` is err .. not the most elegant thing ... but at least you get to know what failed.

Agreed. The null return solution is not nice at all.

> I must confess that I do not know at this point if using exceptions is a problem for you or not. It seems that you can use them. If I am correct please consider using `StatusOr<T>::value()`, as in:

I have no problem with exceptions and prefer them to cumbersome error objects but a lot of people have. Notably game developers. Unreal has them disabled per default and you need to actively enable them. But not everybody has that choice. I personally like the approach other libraries have of offering two versions for such methods. One that throws an exception and one that takes an error object reference that will be set on error. Some Boost libraries do that. Filesystem and some others. Problem is, that doesn't work with constructors and is a bit chunky towards templates.

> ```c++
> StatusOr<Foo> G();
> 
> void F() {
>   Foo v = G().value();
> }
> ```

Same effect sadly. I will try with one of my own types but see if I get C++14 to work first. My hunch is that it has something to do with the optional type prohibiting constructors and destructors because both the error messages I had to disable and the ones I'm getting when using C++14 point in that direction. The one with C++14 even has the method value() you refer to in it. It must be something along those lines...

Thanks for your input and for taking my criticism so nicely.

Moose

**Reply by coryan** [reply]
> > Which portions of `google-cloud-cpp` are you interested in using? If (and based on the discussion below this sounds plausible) you only want GCS, then you could drop the dependencies on gRPC and protobuf. The magic incantation is:
> > ```
> > cmake -DGOOGLE_CLOUD_CPP_ENABLE=storage -S . -B ...
> > ```
> 
> Sadly, this may not be an option. What I require is (at least now) storage, pubsub and ideally logging even though I came to realize that logging is not really available right now. Looks like you have it planned though. Once I understood this I changed my plan to integrate storage first, then pubsub and then wait for logging to be integrated.

Ack.

> > We know this is a problem, and it is something we are constantly working on. The next release might help, building all the dependencies becomes:
✂️  -- some details -- ✂️ 
> 
> Well that looks interesting. Just bear in mind that package managers may not be the silver bullet. I shy away from them when I can because they are forced to make a lot of assumptions about your environment.

Ack.

> Unreal is a good example: They build everything with /MD. Which is great, but unexpectedly, this is also the case for debug builds (as apposed to /MDd). I call that "fake debug". No matter what and how you build, you're always using release runtime. Their debug builds are actually release builds with compiler optimization turned off and debug info turned on. Works just as well as normal debug builds except you don't need debug library dependencies and you can freely mix your builds. I found that to be so great that I use this everywhere instead of bothering with "real" debug.
> Package managers though have no idea about this and often will require twin builds of release and debug because it is written. That's just one example. If your solution you describe will produce a build with vcpkg I will certainly give it a try.

Your point about package managers not being a panacea is well taken. Having said that, the folks in `vcpkg` may have a plan for Unreal, or maybe one can create custom triplet that sets all the right flags.

> > That should download and compile all the dependencies for you. It will not solve the DLL problems, for the reasons in #5849.
> 
> Ahh, that looks familiar. I used to have that problem with protobuf on Windows as well. As far as I knw there was a protoc option that allowed you to inject declspec decorators of your choice into the generated headers. Isn't that what you need here?

Yeah.  It is more than just the declspec decorator, you have to add flags to auto-include the headers that define these decorators. And you need one per library. For "reasons" we have a lot of proto libraries.

> This is Visual Studio 2019, latest patch. Though I don't believe it would occur in a regular VS solution. Unreal only uses VS as some sort of an editor but the actual build system used underneath is UBT (Unreal Build Tool) which does a lot of magick on top and then calls the MSVC compiler. Somewhere in there is probably the reason for this.

Oh joy...

✂️  -- some details -- ✂️ 
> 
> There's an interesting resource here: [akrzemi1/Optional#57](https://github.com/akrzemi1/Optional/issues/57)
> 
> It looks to me as if this is not unheard of when using optional types in this context. How I could fix this is over my head though. I am currently toying around with setting various language options such as C++14. The fix above only worked when using 17. Using 14 so far I only got a compile error "\GoogleCloud_SDK\include\absl/meta/type_traits.h(298): error C4596: 'value': illegal qualified name in member declaration" which I couldn't fix yet.
> 
> > Okay... Can you try a simpler operation first, like `GetObjectMetadata` or `GetBucketMetadata`?
> 
> I will try that and update here. Alas, I doubt it as the problem seems to occur when I dereference the StatusOr, before the actual call happens. At least this is what the debugger says.

Ack.

✂️  -- some details -- ✂️ 
> I do check it. Here's a snippet of my code:
> 

> ```c++
> gc::StatusOr<gcs::Client> client = gcs::Client::CreateDefaultClient();
> if (!client) {
>    // handle error using client.status().message()
>    return;
> }
> 
> gc::StatusOr<gcs::ObjectMetadata> object_metadata = client->InsertObject(bucket_name, object_key, mydata, gcs::ContentType(content_type));
> 			
> // never get here. absl Exception is thrown
> ```

Ah, I think it is the `gcs::ContentType`, that holds an optional inside. If it was not initialized properly then who knows what will happen.

✂️  -- some details -- ✂️ 

> > Were you able to compile and test any of the examples with your setup?
> 
> No, but I haven't tried. I fear this will work outside Unreal and send me on a wild goose chase nowhere but you're right. I should probably give it a go to verify that the build itself is OK.

After what you said about UBT, yes.

> > I must confess that I do not know at this point if using exceptions is a problem for you or not. It seems that you can use them. If I am correct please consider using `StatusOr<T>::value()`, as in:
> 
> I have no problem with exceptions and prefer them to cumbersome error objects but a lot of people have. Notably game developers. Unreal has them disabled per default and you need to actively enable them. But not everybody has that choice. I personally like the approach other libraries have of offering two versions for such methods. One that throws an exception and one that takes an error object reference that will be set on error. Some Boost libraries do that. Filesystem and some others. Problem is, that doesn't work with constructors and is a bit chunky towards templates.

Yup.

> Same effect sadly. I will try with one of my own types but see if I get C++14 to work first. My hunch is that it has something to do with the optional type prohibiting constructors and destructors because both the error messages I had to disable and the ones I'm getting when using C++14 point in that direction. The one with C++14 even has the method value() you refer to in it. It must be something along those lines...

Is it possible to convince UBT to use C++17?  If it is, you can get Abseil always use `std::optional<>`, you need to tweak one of the header files:

https://github.com/abseil/abseil-cpp/blob/master/absl/base/options.h

If you set that, then we will use `std::optional<>`, and maybe we can avoid all these problems.

You would need to do something like:

```shell
sed -i 's/^#define ABSL_OPTION_USE_\(.*\) 2/#define ABSL_OPTION_USE_\1 1/' "absl/base/options.h"
```

PS: If I missed a topic above, please ping me again, I got distracted and my have lost the thread.

**Reply by MrMoose** [reply]
> > This is Visual Studio 2019, latest patch. Though I don't believe it would occur in a regular VS solution. Unreal only uses VS as some sort of an editor but the actual build system used underneath is UBT (Unreal Build Tool) which does a lot of magick on top and then calls the MSVC compiler. Somewhere in there is probably the reason for this.
> 
> Oh joy...

I know. But to be fair, it's quite an interesting system. Your build system is actually written in C#. By you. And then assembled into the actual build system, custom for each build. At least that's the way I understand it. Sort of like if CMake scripts were compiled into an executable build system rather than being interpreted by CMake. It's weird but you get quite a bit out of it when you finally manage to tame it. It manages compiler settings, dependencies among modules, plugins, all kinds of stuff really that would be totally unmanageable in a project the size of an Unreal project.
It just takes a few years to get the hang of it and it has an unfortunate tradition of introducing undocumented breaking changes between releases all the time. When you upgrade an Engine version it often feels like 20% of the time goes into code changes and the rest goes right into figuring out what broke the damn build system again. ;-)

> Is it possible to convince UBT to use C++17? If it is, you can get Abseil always use `std::optional<>`, you need to tweak one of the header files:
> 

Well, that was an interesting suggestion. Yes, UBT can use C++ 17 and in my experiments I was actually using it. When I wrote this I was in the process of downgrading to 14 because I was out of other ideas. Now instead I went to check out all my dependencies and make sure to rebuild them with C++17 explicitly set (abseil and grpc mainly). Doing so I noticed that I couldn't build grpc against abseil unless I explicitly set the option in abseil. 
Got a linker error to exactly the exception type that came up in my original post. Hardly a coincidence. So I went further and rebuild everything, making sure C++17 all the way.
And believe it or not, the call now goes through!

It still doesn't work as I now get this one here: https://github.com/googleapis/google-cloud-cpp/issues/3770 but nevertheless I feel that this is a big step.
When I patch `curl_handle_factory.cc` and disable peer verification in line 33....

```C++
curl_easy_setopt(handle, CURLOPT_SSL_VERIFYPEER, 0L);
```

... it works! I can upload a small file to storage. I believe I can live with the missing verification for the time being. Tried all kinds of tricks to make libcurl find the ca bundle but everything failed.
Thanks for that advise. C++ 17 for the win here. Now I only need to figure out how to prevent CURL from checking the server cert or figure out where to put curl-config but this appears doable. One step further!

Thanks again!

Moose

**Reply by coryan** [reply]
✂️  ---- ✂️ 
> Got a linker error to exactly the exception type that came up in my original post. Hardly a coincidence. So I went further and rebuild everything, making sure C++17 all the way.
> And believe it or not, the call now goes through!

🎉  yay!

> It still doesn't work as I now get this one here: #3770 but nevertheless I feel that this is a big step.
> When I patch `curl_handle_factory.cc` and disable peer verification in line 33....
> 
> ```c++
> curl_easy_setopt(handle, CURLOPT_SSL_VERIFYPEER, 0L);
> ```
> 
> ... it works! I can upload a small file to storage. I believe I can live with the missing verification for the time being. Tried all kinds of tricks to make libcurl find the ca bundle but everything failed.

Maybe consider downloading Google's `roots.pem` file as described in the quickstarts:

https://github.com/googleapis/google-cloud-cpp/tree/master/google/cloud/bigtable/quickstart#windows

you can then point the GCS library to use it:

https://github.com/googleapis/google-cloud-cpp/blob/98139ba3be634cd12615a753e5f5965c51ee5d8c/google/cloud/storage/client_options.h#L290

Best of luck, let us know if we can make this easier.

**Reply by coryan** [reply]
> > It still doesn't work as I now get this one here: #3770 but nevertheless I feel that this is a big step.
> > When I patch `curl_handle_factory.cc` and disable peer verification in line 33....
> > ```c++
> > curl_easy_setopt(handle, CURLOPT_SSL_VERIFYPEER, 0L);
> > ```
> > 
> > 
> > ... it works! I can upload a small file to storage. I believe I can live with the missing verification for the time being. Tried all kinds of tricks to make libcurl find the ca bundle but everything failed.
> 
> Maybe consider downloading Google's `roots.pem` file as described in the quickstarts:

This subject came up in #6212.  Depending on how you build libcurl it can use Schannel (and the native root of trust), or OpenSSL (and then you need settings to find the root of trust, even to use the native one).  I thought it would be worth mentioning here.

**Reply by MrMoose** [reply]
> This subject came up in #6212. Depending on how you build libcurl it can use Schannel (and the native root of trust), or OpenSSL (and then you need settings to find the root of trust, even to use the native one). I thought it would be worth mentioning here.

Thanks for mentioning this! I wouldn't have noticed. As I commented in #6212, building libcurl with Schannel instead of OpenSSL and SSH2 fixes the issue and I can now work with an unpatched lib.

Having users download certificates was not really a viable option to me. It should be working with the 'regular' certificate bundles being installed on most machines.

**Reply by coryan** [reply]
> Having users download certificates was not really a viable option to me. It should be working with the 'regular' certificate bundles being installed on most machines.

gRPC may be a challenge:

https://github.com/grpc/grpc/issues/16571

**Reply by MrMoose** [reply]
> gRPC may be a challenge:
> 
> [grpc/grpc#16571](https://github.com/grpc/grpc/issues/16571)

Eew! I guess I will encounter that very soon then as I am currently working on Pubsub. Haven't tried it out yet. Well, seems like there's no way around then. Thanks for the hint. I guess when I have to use those root certs anyway, I might as well give a curl/openssl build another go and see if that fixes it in the same way.

**Reply by MrMoose** [reply]
A little update on that one in case it is of interest.

Meanwhile I have got a basic impl of pubsub running but unexpectedly without running into https://github.com/grpc/grpc/issues/16571 . I don't know the reason for this. I have no manually downloaded root.pem cert bundle anywhere and yet it still seems to work OK.

**Reply by coryan** [reply]
Good to hear that.  My guess (emphasis on guess) is that `grpc` installed its own `roots.pem`:

https://github.com/grpc/grpc/tree/master/etc

using this rule:

https://github.com/grpc/grpc/blob/8489a6050b64a9c135aaac1c4ccb59b9d5dfe076/CMakeLists.txt#L15517-L15518

I understand such files should be periodically updated, but maybe this will work for you.

**Reply by MrMoose** [reply]
Well, I just had a look at this and you are right, this file is installed. Yet my fears that the path to it is somehow hardcoded in gRPC were unfounded. If I remove it after the build it still works.

It would appear as if either the file's contents are somehow linked into the binary gRPC libs or the file it not needed at runtime after all.
